Simpler spawner/tester

LuKks opened the pull request:
This is just a demo on how we could simplify the spawner to keep all within the same file

LuKks committed:
simpler spawner

LuKks committed:
simpler spawner

LuKks committed:
+

LuKks committed:
+

LuKks committed:
fix helpers path

LuKks committed:
fix helpers path

LuKks committed:
add tap-parser package

LuKks committed:
add tap-parser package

LuKks committed:
+

LuKks committed:
+

LuKks committed:
simpler spawn

LuKks committed:
simpler spawn

LuKks committed:
Update index.js

LuKks committed:
Update index.js

LuKks committed:
remove tap-parser

LuKks committed:
remove tap-parser

LuKks committed:
using mafintosh way of standardize tap

LuKks committed:
using mafintosh way of standardize tap

LuKks committed:
+

LuKks committed:
+

mafintosh committed:
exit with 1 if assert fails

LuKks committed:
improve standardize tap

LuKks committed:
improve standardize tap

LuKks committed:
simpler comparison using tester/spawner

LuKks committed:
simpler comparison using tester/spawner

mafintosh commented on test/plan.js:
```diff
@@ -0,0 +1,59 @@
+    stack: |
+    ...
+    not ok 2 - plan must be positive
+    1..2
```
use the original indented output for the expected output

LuKks committed:
remove t.pass()

LuKks committed:
remove t.pass()

LuKks committed:
+

LuKks committed:
+

LuKks committed:
using original output as expected

LuKks committed:
using original output as expected

LuKks committed:
remove child var as there is a timeout

LuKks committed:
remove child var as there is a timeout

LuKks committed:
fix brittle pkg path in helper

LuKks committed:
fix brittle pkg path in helper

LuKks committed:
capture exitCode

LuKks committed:
capture exitCode

LuKks committed:
Update index.js

LuKks committed:
Update index.js

LuKks committed:
Update plan.js

LuKks committed:
Update plan.js

LuKks committed:
add exit code comparison and reuse more code

LuKks committed:
add exit code comparison and reuse more code

LuKks committed:
Update index.js

LuKks committed:
Update index.js

LuKks committed:
match exitCode on all cases

LuKks committed:
match exitCode on all cases

LuKks committed:
Update plan.js

LuKks committed:
Update plan.js

LuKks committed:
fix spawner with async functions

LuKks committed:
fix spawner with async functions

LuKks committed:
add tapexp

LuKks committed:
add tapexp

LuKks committed:
add debug

LuKks committed:
add debug

LuKks committed:
Create after-end-assert.js

LuKks committed:
Create after-end-assert.js

LuKks committed:
add chalk (temporarily)

LuKks committed:
add chalk (temporarily)

LuKks committed:
better logging

LuKks committed:
better logging

LuKks committed:
Update after-end-assert.js

LuKks committed:
Update after-end-assert.js

LuKks committed:
Update index.js

LuKks committed:
Update index.js

LuKks commented:
I pushed some changes that fix a scenario with `spawner` + `async function`, I added `tapexp` which is the `expected` variable after being standardized.

However, I added `after-end-assert.js` but I noticed there is some differences between `tapout` and `tapexp`, so it doesn't pass:
![image](https://user-images.githubusercontent.com/12686176/184055080-6f027149-f3d5-493c-a848-9d6c8e328e8b.png)

The `stdout` of the inverted test doesn't return the first line like `TAP version 13`.

Now I added some colors so it's easier to read between all the random tap texts. I will remove `chalk` lib after reaching some stable point.

Previous brittle `t.end()` returned a `Promise`, and next brittle doesn't. I like that it doesn't but that is a breaking change as the rare `t.end().then(...)` is not valid anymore, although all tests done with `await t.end()` should mostly continue working.

Is it okay to have classic + inverted (regarding the same test) in a single file? Like the one I pushed: `after-end-assert.js`.
If you still prefer a single test per file, then it could be a `test` folder with test files + `classic` folder + `inverted` folder.

LuKks commented:
It would be cool to try merge this, so it doesn't become a too crazy PR and I can continue working on top with smaller changes, i.e. for now we could just deal with passing `t` to `tester`/`spawner`, later I could improve that. Improvements to `standardizeTap` function, adding more tests over multiple PRs, etc.

mafintosh commented on test/helpers/index.js:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  if (stderr) {
+    throw stderr // + not sure about this
```
throw new Error(stderr)

mafintosh commented on test/helpers/index.js:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  if (error) {
+    throw error // + not sure about this
```
makes sense, only think this will happen if node can't spawn or something

mafintosh committed:
timeout makes the test throw

mafintosh committed:
missing default fix

mafintosh committed:
3.0.0-alpha.2

LuKks committed:
Update index.js

LuKks committed:
Update index.js

LuKks committed:
fix standardizeTap

LuKks committed:
fix standardizeTap

LuKks committed:
disable print

LuKks committed:
disable print

LuKks committed:
not using itself anymore to test

LuKks committed:
not using itself anymore to test

LuKks committed:
remove fixtures folder

LuKks committed:
remove fixtures folder

LuKks committed:
Update index.js

LuKks committed:
Update index.js

LuKks committed:
set exitCode 1 when there are errors

LuKks committed:
set exitCode 1 when there are errors

LuKks committed:
better logging of errors

LuKks committed:
better logging of errors

LuKks committed:
Update index.js

LuKks committed:
Update index.js

mafintosh commented on test/after-end-assert.mjs:
```diff
@@ -0,0 +1,50 @@
+
+  # not ok
+  `,
+  { exitCode: 0 }
```
this should be exitCode 1

mafintosh replied:
ie, the test is failing

mafintosh commented on test/after-end-assert.mjs:
```diff
@@ -0,0 +1,50 @@
+
+  # not ok
+  `,
+  { exitCode: 0 }
```
exitCode 1

mafintosh commented on test/helpers/index.js:
```diff
@@ -90,6 +104,7 @@ function functionToString (func, opts = {}) {
 }
 
 function print (name, color, str) {
+  return
```
meant to be here?

LuKks replied:
yes, I would like to add some debug flag later if that's okay, all the printing can be removed later

mafintosh replied:
ok np

mafintosh commented on test/helpers/index.js:
```diff
@@ -6,42 +6,56 @@ const pkg = JSON.stringify(path.join(__dirname, '..', '..', 'index.js'))
   print('tapexp', 'cyan', tapexp)
 
-  return { exitCode, stdout, tapout, tapexp, stderr }
+  if (errors.length) {
```
set process.exitCode here?

LuKks replied:
I did that, not sure why it wasn't at the moment of the review

LuKks committed:
Merge branch 'add-test-folder' of github.com:LuKks/brittle into add-test-folder

LuKks committed:
avoid not actual child errors

LuKks committed:
Update index.js

LuKks committed:
Update after-end-assert.mjs

mafintosh commented on test/helpers/index.js:
```diff
@@ -0,0 +1,142 @@
+    child.stdout.on('data', (chunk) => stdout.push(chunk))
+    child.stderr.on('data', (chunk) => stderr.push(chunk))
+    child.stdout.on('end', (chunk) => stdout.push(chunk))
+    child.stderr.on('end', (chunk) => stderr.push(chunk))
```
end never has a buffer

mafintosh commented on test/helpers/index.js:
```diff
@@ -0,0 +1,142 @@
+}
+
+function cleanStd (std) {
+  if (!std.length || std[0] === undefined) {
```
if unneeded - just because you append undefined in the `end` listener

mafintosh commented on test/plan.js:
```diff
@@ -0,0 +1,90 @@
+  await tester(t, 'plan must be positive',
+    function (t) {
+      // t.fail('random')
+      // throw new Error('random')
```
remove the commented out lines here

mafintosh commented on test/plan.js:
```diff
@@ -0,0 +1,90 @@
+
+    # not ok
+    `,
+    { exitCode: 0 }
```
should be 1, is it not?

LuKks replied:
ooohhhh, plan.js, no ones care about that one anymore XD I wasn't understanding why you keep seeing the exitCode with zero, I thought you were referring to the .mjs file

LuKks committed:
Update index.js

mafintosh commented on test/plan.js:
```diff
@@ -0,0 +1,90 @@
+
+    # not ok
+    `,
+    { exitCode: 0 }
```
should be 1 also

LuKks committed:
allow multiple errors

LuKks committed:
Delete plan.js

LuKks committed:
Update index.js

LuKks committed:
remove stack and use error message

LuKks committed:
simpler functionToString

LuKks committed:
remove cleanStd

LuKks committed:
remove raw opt

LuKks committed:
Update index.js

LuKks committed:
Create all.mjs

LuKks committed:
Update all.mjs

LuKks committed:
Update index.js

mafintosh committed:
fix test script

LuKks committed:
fix standard

LuKks committed:
using hasOwn instead of "in"

mafintosh committed:
print summary
